sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/SubscriptionPartitionLoader.java

```diff
@@ -47,15 +46,99 @@
   private final SubscriptionPath subscription;
   private final SerializableFunction<TopicPath, Integer> getPartitionCount;
   private final Duration pollDuration;
-  private final boolean terminate;
+  private final SerializableSupplier<Boolean> terminate;
+
+  private class GeneratorFn extends DoFn<Void, SubscriptionPartition> {
+    @ProcessElement
+    public ProcessContinuation processElement(
+        OutputReceiver<SubscriptionPartition> output,
+        RestrictionTracker<Integer, Integer> restrictionTracker) {
+      int previousCount = restrictionTracker.currentRestriction();
+      int newCount = getPartitionCount.apply(topic);
+      if (newCount <= previousCount) {
+        return ProcessContinuation.resume().withResumeDelay(pollDuration);
+      }
+      if (!restrictionTracker.tryClaim(newCount)) {
+        return ProcessContinuation.stop();
+      }
+      Instant ts = previousCount == 0 ? BoundedWindow.TIMESTAMP_MIN_VALUE : getWatermark();
+      for (int i = previousCount; i < newCount; ++i) {
+        output.outputWithTimestamp(SubscriptionPartition.of(subscription, Partition.of(i)), ts);
+      }
+      if (terminate.get()) {
+        return ProcessContinuation.stop();
+      }
+      return ProcessContinuation.resume().withResumeDelay(pollDuration);
+    }
+
+    @GetInitialRestriction
+    public Integer getInitialRestriction() {
+      return 0;
+    }
+
+    @NewTracker
+    public RestrictionTracker<Integer, Integer> newTracker(@Restriction Integer input) {
+      return new RestrictionTracker<Integer, Integer>() {
+        private int position = input;
+
+        @Override
+        public boolean tryClaim(Integer newPosition) {
+          checkArgument(newPosition > position);
+          position = newPosition;
+          return true;
+        }
+
+        @Override
+        public Integer currentRestriction() {
+          return position;
+        }
+
+        @Override
+        public @Nullable SplitResult<Integer> trySplit(double fractionOfRemainder) {
+          return null;
+        }
+
+        @Override
+        public void checkDone() throws IllegalStateException {}
```

Comment:
this would have made sense to have embedded the `terminate` function in the restriction tracker and ensured that it returned true.

Reply:
I can't since I need to terminate regardless of if there is new data. Also, terminate is only used for testing.